### ISSUE TITLE
autoreload problems when script using multiprocessing is %run

### ISSUE THREAD

**Issue by azyr** [ipython/ipython#7830]
I recently updated to ipython 2.4.1 from some pretty old version (using pip). I have python 3.4.0 and newly upgraded Linux Mint (Cinnamon). Autoreload magic seems to be broken now.

Now when I run:
```
%load_ext autoreload
%autoreload 2
run myscript.py
```
I get the following error:
```
Error in callback <bound method AutoreloadMagics.post_execute_hook of <autoreload.AutoreloadMagics object at 0x7f60a2fdf518>> (for post_execute):
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/home/seb/.virtualenvs/p3/lib/python3.4/site-packages/IPython/extensions/autoreload.py in post_execute_hook(self)
    522         newly_loaded_modules = set(sys.modules) - self.loaded_modules
    523         for modname in newly_loaded_modules:
--> 524             _, pymtime = self._reloader.filename_and_mtime(sys.modules[modname])
    525             if pymtime is not None:
    526                 self._reloader.modules_mtimes[modname] = pymtime

/home/seb/.virtualenvs/p3/lib/python3.4/site-packages/IPython/extensions/autoreload.py in filename_and_mtime(self, module)
    187             return None, None
    188 
--> 189         if module.__name__ == '__main__':
    190             # we cannot reload(__main__)
    191             return None, None

AttributeError: 'module' object has no attribute '__name__'
```

**Comment by takluyver** [comment]
Can you stick a `print(modname)` in there (just before line 524) to see what module it's choking on?

**Comment by azyr** [comment]
takluyver: I tried that.

__mp_main__ is the modname giving problems.

**Comment by takluyver** [comment]
Does your script use multiprocessing? I guess that's what mp stands for.

**Comment by azyr** [comment]
takluyver: indeed it does. This problem seems to occur only when using ```run``` to scripts that use ```multiprocessing```

**Comment by azyr** [comment]
takluyver: Indeed that seems to be the problem. I get that error when I ```%run``` a program like this:

```python
import multiprocessing

if __name__ == "__main__":

    print("hello")
```

**Comment by takluyver** [comment]
OK, I've retitled the issue to reflect that.

**Comment by memeplex** [comment]
One relevant offender from the point of view of ipython is sklearn. Just importing, say, `linear_model`, drags in the joblib external, which sets the main module to `__mp_main__`. A minimal example is:

```
import sklearn.externals.joblib
```
or

```
import sklearn.linear_model
```

Take a look at this ipython session:

```
In [1]: import sys

In [2]: print(id(sys.modules['__main__']))
140342138805448

In [3]: print(id(sys.modules['__mp_main__']))
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-3-94ef375effaa> in <module>()
----> 1 print(id(sys.modules['__mp_main__']))

KeyError: '__mp_main__'

In [4]: %run /tmp/test.py
__main__
<module '__main__' from '/tmp/test.py'>
<module '__main__' from '/tmp/test.py'>
140342086547560
140342086547560
__main__
__main__

In [5]: print(id(sys.modules['__main__']))
140342138805448

In [6]: print(id(sys.modules['__mp_main__']))
140342086547560

In [7]: 
Do you really want to exit ([y]/n)? y
[carlos@carlos ~]$ ipython
Python 3.4.3 (default, Feb 26 2015, 23:01:07) 
Type "copyright", "credits" or "license" for more information.

IPython 3.0.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import sys

In [2]: print(id(sys.modules['__main__']))
140213167095144

In [3]: print(id(sys.modules['__mp_main__']))
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-3-94ef375effaa> in <module>()
----> 1 print(id(sys.modules['__mp_main__']))

KeyError: '__mp_main__'

In [4]: %run /tmp/test.py
__main__
140213114820472
140213114820472
__main__
__main__

In [5]: print(id(sys.modules['__main__']))
140213167095144

In [6]: print(id(sys.modules['__mp_main__']))
140213114820472

In [7]: print(sys.modules['__mp_main__'].__name__)
AttributeError: 'module' object has no attribute '__name__'
```

where `/tmp/test.py` is defined as:

```
import sklearn.linear_model

print(__name__)

import sys

print(id(sys.modules['__main__']))
print(id(sys.modules['__mp_main__']))

print(sys.modules['__main__'].__name__)
print(sys.modules['__mp_main__'].__name__)
```

The `__main__` module for the `/tmp/test.py` script is aliased as `__mp_main__` in `multiprocessing.__init__` (see the `__init__.py` file for multiprocessing). But this `__main__` module is not the same than the ipython's one (note the different ids in the output above). Still the `__mp_main__` alias remains in `sys.modules` after the `%run`. But this module now has no `__name__` (nota bene: it had one inside the script, namely `__main__`). So:

1. It seems desirable that ipython cleaned the `__mp_main__` module up after `%run` (and not just replace the `__main__` module for the script with the ipython's one).
2. It's not clear why the `__name__` attribute is disappearing. Maybe this has to do with some post-`%run` cleanup ipython is already doing.

**Comment by memeplex** [comment]
@takluyver what do you think of the above solution (clean up `__mp_main__` after `%run`)? It's basically what I'm doing manually now. Seriously, using ipython with sklearn has become a bit of a PITA because of this issue.

**Comment by takluyver** [comment]
I'd be inclined to just make autoreload skip over modules called `__mp_main__`. This is only an issue when using autoreload, right? It's not a problem with IPython in general?

**Comment by memeplex** [comment]
I'm unable to reproduce this anymore. I don't see any explicit reference to `__mp_main__` in `autoreload.py`, so who knows what has changed meanwhile... maybe you?

Could you check whether this is somehow fixed or somehow not?

**Comment by azyr** [comment]
IPython 4.0.0 still has the same bug. 

To reproduce follow these steps:

Write script mp_bug.py:
```python
import multiprocessing

if __name__ == "__main__":

    print("hello")
```
And execute:
```
ipython3
> %load_ext autoreload
> %autoreload 2
> run mp_bug.py
```

**Comment by takluyver** [comment]
I think this is fixed for 4.1 by #8568
